packages/xstate-store/test/react.test.tsx

```diff
@@ -266,3 +266,35 @@ it('useActorRef (@xstate/react) should work', () => {
 
   expect(countDiv.textContent).toEqual('1');
 });
+
+it('useSelector should not infinite loop', () => {
```

Comment:
this is a known caveat listed in [the React docs](https://react.dev/reference/react/useSyncExternalStore#caveats):
> The store snapshot returned by getSnapshot must be immutable. If the underlying store has mutable data, return a new immutable snapshot if the data has changed. Otherwise, return a cached last snapshot.

Is it worth "patching" the problem on our side if this is just how things are "supposed" to work?

Reply:
Wouldn't zustand v5 have the same problem? The solution here is heavily inspired by how they do it: 

https://github.com/pmndrs/zustand/blob/b93e11681d7b70c60dc99ae5296216a066f466b7/src/react/shallow.ts#L4

I think the solution here would be to have the user provide a different `compare` method like `shallow`. Otherwise, yes, returning new instances will have this impact.

Reply:
To clarify, the problem here isn't that it rerenders every time; that's expected and a good use-case for providing a different `compare` method.

The problem is that it will cause an infinite loop:

> Error: Maximum update depth exceeded. This can happen when a component repeatedly calls setState inside componentWillUpdate or componentDidUpdate. React limits the number of nested updates to prevent infinite loops.

Reply:
When this happens with the current code React even prints a warning about it:
> The result of getSnapshot should be cached to avoid an infinite loop